src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -438,6 +445,31 @@ public void close()
      */
     public static class ScoreOrderedResultRetriever extends AbstractIterator<UnfilteredRowIterator> implements UnfilteredPartitionIterator
     {
+        private static final LocalAwareExecutorService PARALLEL_EXECUTOR = getExecutor();
+
+        /**
+         * Executor to use for parallel index reads.
+         * Defined by -Dcassandra.index_read.parallele=true/false, true by default.
+         *
+         * INDEX_READ uses 2 * cpus threads by default but can be overridden with -Dcassandra.index_read.parallel_thread_num=<value>
+         *
+         * @return stage to use, default INDEX_READ
+         */
+        private static LocalAwareExecutorService getExecutor()
+        {
+            boolean isParallel = CassandraRelevantProperties.USE_PARALLEL_INDEX_READ.getBoolean();
+
+            if (isParallel)
+            {
+                int numThreads = CassandraRelevantProperties.PARALLEL_INDEX_READ_NUM_THREADS.isPresent()
+                                 ? CassandraRelevantProperties.PARALLEL_INDEX_READ_NUM_THREADS.getInt()
+                                 : FBUtilities.getAvailableProcessors() * 2;
+                return SharedExecutorPool.SHARED.newExecutor(numThreads, maximumPoolSize -> {}, "request", "IndexParallelRead");
+            }
+            else
+                return ImmediateExecutor.INSTANCE;
```

Comment:
Nit: if we are to follow ASF style rules (we might not), use brackets if the other branch has brackets:
```suggestion
            {
                return ImmediateExecutor.INSTANCE;
            }
```